### PR TITLE
Initial implementation of Difference

### PR DESCRIPTION
**Pull request by Daroslaw**
This is an implementation of Matlab's diff function specified here:
https://www.mathworks.com/help/matlab/ref/diff.html

The main difference is support for only 1D data for now, for `mspeaks` needs.

I attempted to perform as least copies and memory reallocations as possible and I'm open for suggestions how to improve this code performance-wise. My first thought is to give up on `minus()` function and perform all the subtractions in an inner loop. Maybe I'll do some benchmarking on it soon.

### CODE REVIEW

File: src/Spectre.libStatistics/Math.h [hunk 1]
```diff
@@ -380,4 +380,30 @@ std::vector<DataType> build(size_t size, Generator generator)
...
 }
 
+/// <summary>
+/// Calculates differences between adjacent elements of given container.
+/// </summary>
+/// <param name="data">The container.</param>
+/// <param name="order">Order of difference operation (how many times the difference is executed).</param>
+/// <returns>Vector of calculated differences.</returns>
+ template <class DataType>
+ std::vector<DataType> diff(gsl::span<const DataType> data, uint16_t order = 1)
```
Comment (Sand3r-) [inline src/Spectre.libStatistics/Math.h]: I'd rename that to Differenciate

Comment (gmrukwa) [inline src/Spectre.libStatistics/Math.h]: `std::vector<std::remove_const<DataType>::type> diff(gsl::span<DataType> data, uint16_t order = 1)`

File: src/Spectre.libStatistics/Math.h [hunk 2]
```diff
@@ -380,4 +380,30 @@ std::vector<DataType> build(size_t size, Generator generator)
...
+ {
+    size_t length = data.length();
+    std::vector<DataType> result = std::vector<std::remove_const<DataType>::type>(data.begin(), data.end());
```
Comment (gmrukwa) [inline src/Spectre.libStatistics/Math.h]: `std::vector<std::remove_const<DataType>::type> result(data.begin(), data.end());`

File: src/Spectre.libStatistics/Math.h [hunk 3]
```diff
@@ -380,4 +380,30 @@ std::vector<DataType> build(size_t size, Generator generator)
...
+
+    if (order == 0 || length < 2 || order >= length)
```
Comment (gmrukwa) [inline src/Spectre.libStatistics/Math.h]: As I agree with first part )`order == 0`, the later are strange to me. Why don't we do usual validation with exception on failure? Or at least something in the documentation, why the behavior is such...

File: src/Spectre.libStatistics/Math.h (hunk 2 above)
Comment (Sand3r-) [inline src/Spectre.libStatistics/Math.h]: Aren't you invoking vector copy here?
Think it'd be safer to limit it down to:
```C++
std::vector<std::remove_const<DataType>::type> result(data.begin(), data.end());
```

File: src/Spectre.libStatistics/Math.h [hunk 4]
```diff
@@ -380,4 +380,30 @@ std::vector<DataType> build(size_t size, Generator generator)
     return result;
 }
 
+/// <summary>
+/// Calculates differences between adjacent elements of given container.
```
Comment (gmrukwa) [inline src/Spectre.libStatistics/Math.h]: Please add information about the source of all assumptions that should be taken into account w.r.t. this function, i.e. page with MATLAB specification of their `diff` function behaviour. It'll be crucial to know, that e.g. output size is input size - 1, while it's not so explicit in tests.

File: src/Spectre.libStatistics/Math.h [hunk 5]
```diff
@@ -380,4 +380,30 @@ std::vector<DataType> build(size_t size, Generator generator)
...
+    std::vector<DataType> result = std::vector<std::remove_const<DataType>::type>(data.begin(), data.end());
+
+    if (order == 0 || length < 2 || order >= length)
+        return result;
+
+    for(auto i = 0u; i < order; ++i)
+    {
+        gsl::span<const DataType> lhs = gsl::as_span(result.data() + 1, result.size() - 1);
+        gsl::span<const DataType> rhs = gsl::as_span(result.data(), result.size() - 1);
+        memcpy(result.data(), std::vector<DataType>(minus(lhs, rhs)).data(), (result.size() - 1) * sizeof(DataType));
```
Comment (gmrukwa) [inline src/Spectre.libStatistics/Math.h]: We're allocating the memory either way. Why don't we simply do `result = minus(lhs, rhs)`? It would exercise `move` operator, without copying and resizes.

Comment (Daroslaw) [reply src/Spectre.libStatistics/Math.h]: Nuhuh, such assignment changes the position in memory of `result`. This approach preserves it.

File: src/Spectre.libStatistics/Math.h (hunk 3 above)
Comment (Daroslaw) [reply src/Spectre.libStatistics/Math.h]: Alright, I'll do some exceptions.

File: src/Spectre.libStatistics/Math.h (hunk 5 above)
Comment (gmrukwa) [reply src/Spectre.libStatistics/Math.h]: But we're allocating the memory in `minus` either way, while constant position of `result` in memory gives us no gain. Am I wrong?

File: src/Spectre.libStatistics/Math.h (hunk 3 above)
Comment (Sand3r-) [inline src/Spectre.libStatistics/Math.h]: This could be optimized to
```C++
uint16_t skip = 0 - !(length < 2 || order >= length);
order = order & skip;
```
That way order will be equal to 0 whenever the condition fails, meaning the loop wouldn't execute. 
However, it'd be nice to explain in comment what does that operation do.

File: src/Spectre.libStatistics.Tests/MathTest.cpp [hunk 6]
```diff
@@ -218,4 +218,11 @@ TEST(AbsTest, returns_absolute_value_of_each_element)
     const auto absolutes = abs(gsl::as_span(data));
     EXPECT_THAT(absolutes, ContainerEq(Data{ 1,2,3,4 }));
 }
+
+TEST(DiffTest, calculates_differential)
```
Comment (Sand3r-) [inline src/Spectre.libStatistics.Tests/MathTest.cpp]: Would be nice to add tests when an empty vector is passed, or when the length of vector is lower than the requested order.

File: src/Spectre.libStatistics/Math.h (hunk 3 above)
Comment (Sand3r-) [reply src/Spectre.libStatistics/Math.h]: I wish we could ```zjeść ciastko i mieć ciastko```, but no exceptions can be thrown without branches :worried:.

File: src/Spectre.libStatistics/Math.h (hunk 5 above)
Comment (Daroslaw) [reply src/Spectre.libStatistics/Math.h]: That is why I considered doing it in some other way than using `minus`. It is here mostly just for the same of using already created functionality. I will tinker with `transform`.

File: src/Spectre.libStatistics/Math.h (hunk 1 above)
Comment (Daroslaw) [reply src/Spectre.libStatistics/Math.h]: What? Why would you want to change the signature of the function to return type that differs from every single return type of any mathematical function in here? Please elaborate.

File: src/Spectre.libStatistics/Math.h (hunk 3 above)
Comment (Daroslaw) [reply src/Spectre.libStatistics/Math.h]: The branching here isn't that painful as it is only a single check, then performing branchless algorithm.

File: src/Spectre.libStatistics/Math.h (hunk 4 above)
Comment (Daroslaw) [reply src/Spectre.libStatistics/Math.h]: Done

File: src/Spectre.libStatistics.Tests/MathTest.cpp (hunk 6 above)
Comment (Daroslaw) [reply src/Spectre.libStatistics.Tests/MathTest.cpp]: Done

File: src/Spectre.libStatistics/Math.h (hunk 3 above)
Comment (gmrukwa) [reply src/Spectre.libStatistics/Math.h]: Compiler predicts branches preety well :)

File: src/Spectre.libStatistics/Math.h [hunk 7]
```diff
@@ -380,4 +381,43 @@ std::vector<DataType> build(size_t size, Generator generator)
...
+/// </summary>
+/// <param name="data">The container.</param>
+/// <param name="order">Order of difference operation (how many times the difference is executed).</param>
+/// <returns>Vector of calculated differences, of size: length_of_data - order.</returns>
+template <class DataType>
+std::vector<typename std::remove_const<DataType>::type> differentiate(gsl::span<DataType> data, uint16_t order = 1)
+{
+    size_t length = data.length();
+    if (length < 2)
+        throw std::invalid_argument("Data must have length greater than two in order to perform differentiation.");
```
Comment (gmrukwa) [inline src/Spectre.libStatistics/Math.h]: We have our exception base to differentiate our exceptions of the unhandled ones. We have even `ArgumentOutOfRangeException` or something like this in `spectre::core::exception`. It could be of not too bad use here. (You can specify `data.size()` as bad parameter).

For the future: do not provide user-oriented messages at exception throwing site. Exception should gather the context in such a manner that does not violate DRY and is maintainable. Custom messages at each throw site aren't.

File: src/Spectre.libStatistics/Math.h [hunk 8]
```diff
@@ -380,4 +381,43 @@ std::vector<DataType> build(size_t size, Generator generator)
...
+/// Calculates differences between adjacent elements of given container.
+/// The function is based upon Matlab function diff():
+/// https://www.mathworks.com/help/matlab/ref/diff.html
+/// but it can handle only one dimensional data.
+/// </summary>
+/// <param name="data">The container.</param>
+/// <param name="order">Order of difference operation (how many times the difference is executed).</param>
+/// <returns>Vector of calculated differences, of size: length_of_data - order.</returns>
+template <class DataType>
+std::vector<typename std::remove_const<DataType>::type> differentiate_unsafe(gsl::span<DataType> data, uint16_t order = 1)
```
Comment (Sand3r-) [inline src/Spectre.libStatistics/Math.h]: We'll need to preapre an issue for revising the naming convention in this header file.
